gensim/corpora/dictionary.py

```diff
@@ -145,35 +145,35 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
 
         If `allow_update` is **not** set, this function is `const`, aka read-only.
         """
-        result = {}
-        missing = {}
         if isinstance(document, string_types):
             raise TypeError("doc2bow expects an array of unicode tokens on input, not a single string")
-        document = sorted(utils.to_unicode(token) for token in document)
-        # construct (word, frequency) mapping. in python3 this is done simply
-        # using Counter(), but here i use itertools.groupby() for the job
-        for word_norm, group in itertools.groupby(document):
-            frequency = len(list(group)) # how many times does this word appear in the input document
-            tokenid = self.token2id.get(word_norm, None)
-            if tokenid is None:
-                # first time we see this token (~normalized form)
-                if return_missing:
-                    missing[word_norm] = frequency
-                if not allow_update: # if we aren't allowed to create new tokens, continue with the next unique token
-                    continue
-                tokenid = len(self.token2id)
-                self.token2id[word_norm] = tokenid # new id = number of ids made so far; NOTE this assumes there are no gaps in the id sequence!
-
-            # update how many times a token appeared in the document
-            result[tokenid] = frequency
+
+        # Construct (word, frequency) mapping.
+        counter = defaultdict(int)
+        for w in document:
+            counter[w if isinstance(w, unicode) else unicode(w, 'utf-8')] += 1
+
+        token2id = self.token2id
+        if allow_update or return_missing:
+            missing = dict((w, freq) for w, freq in iteritems(counter)
+                               if w not in token2id)
+            if allow_update:
+                for w in missing:
+                    # new id = number of ids made so far;
+                    # NOTE this assumes there are no gaps in the id sequence!
+                    token2id[w] = len(token2id)
+
+        result = dict((token2id[w], freq) for w, freq in iteritems(counter)
+                  if w in token2id)
 
         if allow_update:
             self.num_docs += 1
-            self.num_pos += len(document)
+            self.num_pos += sum(itervalues(counter))
```

Comment:
Ah never mind, I see now -- Lars' commit introduced a bug where generators no longer worked. Got it.